### ISSUE TITLE
Error handling

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e-testing#32]
While the page is loading we can add a window.onerror handler. What I am not sure of whether we should separate loading errors and js errors. I guess if we have loading errors, then the new page won't be loaded after the unload event.  I am not sure what the abort event covers: https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onabort It is supported by Chrome, so maybe we can use it. I guess I have to add that too every time I load a new page. If we really separate these loading errors from the coding errors, then we need to name them and add them to the api. I don't know yet how I will solve that. If we don't separate them, then a simple onerror will be ok for every error. Sending the error details in param is necessary. Instead of onabort we could have 2 errors types: LoadingError and CodingError. I need to test how this onabort works and what kind of errors can happen while we are loading a page.

**Comment by inf3rno** [comment]
We need an experiment check what happens by loading errors: #45 . After we are done we might need another experiment to check whether any of the event handlers is fired by these loading errors, or what is the best way to detect them. The unload+nexttick might be the best strategy. I won't do a separate issue for that, we can develop what we find into the code.

I am curious whether an abort event or something like that is fired. In theory we can use window.stop to check what happens: https://developer.mozilla.org/en-US/docs/Web/API/Window/stop that will stop resource loading according to the doc, but it does not interfere with page loading. Closing the child window while it is loading might be another option. I am not sure whether I need to do anything about that, it might depend on whether it fires an onabort event somewhere or not.

**Comment by inf3rno** [comment]
Separating the loading errors from the coding errors is a good idea. It is possible to do so, we can follow the loading process and we can add an onerror listener to capture coding errors. The only problem that without the loading handler we won't be able to add that event listener in time. which means we have to drop browser support. otherwise we won't be able to handle errors.

**Comment by inf3rno** [comment]
Another option to use iframes and MessageChannel or postMessage with namespaces. If so we need to add a js file to the injection server. Yet another option to serve the e2e tester on the same origin, but that is not possible with karma.

**Comment by inf3rno** [comment]
In #50 we found the defer section of the Promise polyfill. That means if we use the polyfill and we have no proper defer method for it (MutationObserver, process.nextTick, vertxNext), then Promise.resolve will be delayed. So we need to check it for delay if we use the polyfill, or we need to exclude it if we use the polyfill, since it uses our other defer solutions. So somehow we need to save whether it was polyfilled or not.

**Comment by inf3rno** [comment]
I fixed the polyfilled version, so it won't interfere with defers.

Let's sum up the browser support for defers:

Chrome 68, Opera 54

	after:unload -> Promise.resolve, MutationObserver
	before:loading -> requestIdleCallback 1, postMessage, MessageChannel
	after:complete -> requestIdleCallback, load + defer
	
Firefox 61

	after:unload -> Promise.resolve, MutationObserver
	before:loading -> requestIdleCallback 1, postMessage, MessageChannel, setTimeout 0, requestAnimationFrame
	after:complete -> requestIdleCallback, load + defer
	
Phantom 2.1

	after:unload -> MutationObserver
	before:loading -> -
	after:complete -> postMessage, MessageChannel, requestAnimationFrame, setTimeout 0
	
IE 10, IE 11

	after:unload -> unload
	before:loading -> -
	after:complete -> MutationObserver/script.readyStateChange, MessageChannel, postMessage, requestAnimationFrame, setImmediate, setTimeout 0
	
Edge 42

	after:unload -> Promise.resolve
	before:loading -> -
	after:complete -> MessageChannel, postMessage, requestAnimationFrame, setImmediate, setTimeout 0
	
Safari 10

	after:unload -> Promise.resolve, MutationObserver
	before:loading -> -
	after:complete -> postMessage, MessageChannel, requestAnimationFrame, setTimeout 0

**Comment by inf3rno** [comment]
The after:unload -> unload is an acceptable solution. The only drawback that a script can add an unload event listener, which runs later than our unload listener and which redirects the page or steps back in history. I think that is very rare in a normal browser, so we can use unload in IE. I think we should remove process.nextTick from the defers because we will use only browsers. I know there is a nodejs based browser called Zombie, but that can do e2e tests on its own, so it would not make any sense to support that.

**Comment by inf3rno** [comment]
As we already talked about there is no way to add reliable console logging and error capturing without the before:loading event. For IE 10, IE 11, Edge 42 and Safari 10 we need to add a script to the head of every tested pages. It should contain something like `parent.e2e.fireEvent("loading", {source:window});`. On the long run we'll need an injection server anyways, so this won't be a huge problem I guess. For the injection server the messaging is supported in every browser, so that won't be a problem. As far as I know it can even transfer objects and functions, but I am not sure.

**Comment by inf3rno** [comment]
I think we need to bind to the load event too if we use an injection server. So we won't need defers in the parent window to trigger the complete event in time.

**Comment by inf3rno** [comment]
For Chrome we can use the current alternative method because in Chrome we can use security:disabled, so cross-frame + cross-origin scripting amongst windows will be allowed.

**Comment by inf3rno** [comment]
In our previous plans we have a page for the injection server and send the tests with postMessage to avoid cross-origin checks, but that would not work because the error capturing and the console logging wouldn't be reliable that way.

**Comment by inf3rno** [comment]
There are many anti-iframe solutions and if we use for example a CMS on server it is hard to turn them off. It would be better to avoid using iframes. That means abortion errors will be hard to test. I need to check what I can do about this on different browsers.

**Comment by inf3rno** [comment]
The navigation error and the page error is separated now on an api level. The navigation error must include abortion and maybe 404 errors. The only problem with the latter that we cannot read what response headers were loaded into the page. The only way to go around that is using XHR, but not every request is idempotent, so sending everything twice would be a problem. I think the tests after page load will cover that part because of the missing content.. So for now only abortion will be a navigation error, and we need an experiment to find out what happens when a page cannot be loaded or a server is not reachable without an actual 404 response. Maybe abort covers that part too, maybe not, we need to test it. We need to test whether window.stop() covers abort in every browser.

**Comment by inf3rno** [comment]
I checked with timeout by non-existing pages. 

```js
    load(resourceIdentifier) {
        return Promise.race([
            new Promise((resolve, reject) => {
                this.once("page:complete", () => {
                    resolve(this.window);
                });
                this.once("navigation:error", reject);
                this.window.document.location.href = resourceIdentifier;
            }),
            new Promise((resolve, reject) => {
                setTimeout(() => {
                    console.log(this.window.document.readyState, this.window.document.location.href);
                    reject("navigation:timeout", this.options.navigationTimeout);
                }, this.options.navigationTimeout);
            })
        ]);
    }
```

The timeout is never called. It seems like even non-existing pages are loaded somehow. I'll continue tomorrow.

**Comment by inf3rno** [comment]
The `await windowManager.load("http://www.qqqqqqqqqqq.hu/");` result the following log:

```
LOG: '#1', 'window:open'
LOG: '#1', 'page:unload'
LOG: '#1', 'navigation:request'
LOG: '#1', 'navigation:response', 5
LOG: '#1', 'page:loading', 5
LOG: '#1', 'page:interactive', 26
LOG: '#1', 'page:complete', 27
LOG: '#1', 'window:close'
```

So it can load a non-existent page, which is weird. I have to check the content of the page maybe I can figure out how to detect errors. I have to start writing real tests instead of just experimenting. I want to be complete with 0.1 this week.

**Comment by inf3rno** [comment]
The outerHTML contains some sort of Chromium related template based page. It is not that useful. At least we know what it contains. The location object is a lot more useful:

```
LOG: Location{replace: function () { ... }, assign: function () { ... }, href: 'chrome-error://chromewebdata/', ancestorOrigins: DOMStringList{}, origin: 'null', protocol: 'chrome-erro
r:', host: 'chromewebdata', hostname: 'chromewebdata', port: '', pathname: '/', search: '', hash: '', reload: function reload() { ... }, toString: function toString() { ... }}

```

**Comment by inf3rno** [comment]
I found a polyfill for `window.stop()`. In IE `document.execCommand("Stop");` should work. I am not working currently on any cross-browser stuff, but maybe later this will be needed.

**Comment by inf3rno** [comment]
The interesting thing, that we have to use `window.stop()` when Chrome navigates to an error page and we are in a loading readyState. Apparently only interactive or complete pages can be navigated again by rewriting the location.href. So without stopping the page loading the promise will return in a loading state with the navigation error and the next test will try to use the navigator when it is still busy. We could use a navigator.busy to protect page loading, but first we need to test whether overriding the location.href in the beforeunload, pagehide or unload listeners changes anything.

**Comment by inf3rno** [comment]
Basic error handling is more or less ok. What I need is fixing syntax errors, rejection errors, etc. to get more info from the event handlers. Currently they just say that there was an error, and that's all. I might need to rebuild the stack trace, I am not sure.

**Comment by inf3rno** [comment]
The abortion will be debugged in 0.2.

**Comment by inf3rno** [comment]
This missing error looks like a limitation of Chrome. By undeclared variables it does not pass any arguments. I am not sure I can workaround this. Probably not. I need to check what happens by thrown errors. I hope that works.

**Comment by inf3rno** [comment]
The rejection works properly, only the onerror does not work.

**Comment by inf3rno** [comment]
I need a minimalistic example to decide whether this is a browser error or something different.

x.html
```
<script>
window.opener.childScriptRan = true;
window.addEventListener("error", function (){
	window.opener.console.log("from child", arguments.length, ...arguments);
});
throw new Error("Something bad happened while loading the page.");
</script>
```

loader.html
```
<script>
window.childScriptRan = false;

function assert(condition, message){
	if (!condition)
		throw message;
}

const win = window.open("about:blank", Math.random());
win.addEventListener("unload", function (){
	requestIdleCallback(function (){
		assert(win.document.readyState == "loading", "The document should be in a loading ready state.");
		assert(childScriptRan === false, "The child script shouldn't have run yet.");
		win.addEventListener("error", function (){
			assert(childScriptRan === true, "The child script should have already run.");
			console.log("from parent", arguments.length, ...arguments);
		});
	}, {timeout:1});
});
</script>
```

Surprising, but it works. So 90% that this is not a Chrome error.

**Comment by inf3rno** [comment]
It has a single argument, an error event and the error property of that event contains the cause.

**Comment by inf3rno** [comment]
Ok, this works now. I don't know why it did not list the event as an argument.
